### DISCUSSION TITLE
v0.2 News since 20221107

### DISCUSSION THREAD

**Discussion by dixyes** [swow/swow#136] [Ideas]
（purposed/v0.2.x上发生的事，也是未来的ci分支上会发生的事）

# TL;DR

一波API重构：

* `Buffer`类取消内置偏移，这解决了许多问题
* 重构`HttpParser->execute()`: 现在它返回的是parsed长度而非事件了

一些新功能：

+ 完整的PSR支持
+ IpAddress类：简单的ip地址parser和一些实用api
+ 一些dns解析函数的hook
+ 

Major API refactor:

* Removed internal offset in `Buffer` class, this solves many bugs
* Refactor `HttpParser->execute()`: now it returns parsed length rather than events

Something New:

+ Full PSR compatibility
+ IpAddress class: simple IP address parser and some utility APIs
+ some hook for dns lookup functions
+ 

开发进展记录

## 修复

* 修复`detectShouldKeepAlive()`检查 (#132)
* 修复了许多静态分析警告
* 修复上传文件不完整
* 例行优化代码
* 修复了一些库API参数类型
* 修复了一些情况下URI_TOO_LARGE/HEADER_FIELDS_TOO_LARGE报错

## 增加 删除 修改

+ 库测试覆盖率支持（#131）
+ 添加http parser方法`Http\Parser->getPreviousEvent()` `Http\Parser->getPreviousEventName()` `Http\Parser::getEventNameFor(int $event)
+ `WebSocket::mask()` 和 `WebSocket::unmask()` 工具函数

- 移除 `Socket->sendString()`, 现在可以用`Socket->send()`
- 移除 `WebSocket\Frame`， Websocket现在使用`WebSocket\Header`类加自己的payload

* 重构`HttpParser->execute()`: 现在它返回的是parsed长度而非事件了
* 一些API重命名：`Errno::` `getNameFor(int) => getNameOf(int)` `getReasonPhraseFor(int) => getReasonPhraseOf(int)` `getDescriptionFor(int) => getDescriptionOf(int)`

(记录随时变更)

## Fix

* Fixed `detectShouldKeepAlive()` check (#132)
* Fixed many static analyzers warnings
* Fixed uncompleted uploadedFile data
* Optimize codes as usual
* Fixed some APIs arguments in library
* Fix URI_TOO_LARGE/HEADER_FIELDS_TOO_LARGE at some situation

## Addition Removal Change

+ Library tests coverage (#131)
+ Add new http parser methods `Http\Parser->getPreviousEvent()` `Http\Parser->getPreviousEventName()` `Http\Parser::getEventNameFor(int $event)`
+ `WebSocket::mask()` and `WebSocket::unmask()` utility

- Removed `Socket->sendString()`, use `Socket->send()` instead
- Removed `WebSocket\Frame`, use `WebSocket\Header` class with payload instead
- 

* Refactor `HttpParser->execute()`: now it returns parsed length rather than events
* Some API rename `Errno::` `getNameFor(int) => getNameOf(int)` `getReasonPhraseFor(int) => getReasonPhraseOf(int)` `getDescriptionFor(int) => getDescriptionOf(int)`
* 
(this may change at anytime)

**Comment by AuroraYolo** [comment]
我明明看到邮箱有个多线程的支持，怎么没有了
